Fix map concurrency

gw000 opened the pull request:
Hi, thanks for this great project. The concurrent writes issue from #44  seems to be crashing my pacoloco installation every couple of days. I think this change will fix the problem without adding too much extra complexity: the top-level `time.Time`-valued maps are removed and the times are stored on each `Repo` instead; and the write to `config.Repos` is removed by storing `*Repo` values and giving each one its own mutex.

Please let me know what else i can do to help. Thank you! :slightly_smiling_face:

Focshole commented on config.go:
```diff
@@ -28,15 +33,15 @@ type RefreshPeriod struct {
-	LogTimestamp    bool            `yaml:"set_timestamp_to_logs"`
+	CacheDir        string           `yaml:"cache_dir"`
+	Port            int              `yaml:"port"`
+	Repos           map[string]*Repo `yaml:"repos,omitempty"`
```
Is there a reason to switch to pointers? Is it efficiency related?

gw000 replied:
It's so the repo record can be updated directly, rather than by writing to the config map.

Focshole replied:
Makes sense :+1:

Focshole commented on config_test.go:
```diff
@@ -48,16 +48,16 @@ repos:
 		Prefetch:        &RefreshPeriod{Cron: "0 0 3 * * * *", TTLUnaccessed: 5, TTLUnupdated: 200},
 	}
-	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch")) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch"), cmpopts.IgnoreUnexported(Repo{})) {
```
Well, this will ignore the repo field, right? It is not what I would expect from a test that has to check if configs are matching. I am a bit perplexed from this edit

gw000 replied:
I believe it only ignores the new unexported fields on the `Repo` struct. Apologies if i'm mistaken, i haven't used `cmpopts` before

Focshole replied:
AFAIK it would not check that the Repo pointers addresses match, which is fine. However, the content pointed from the Repo pointers should be checked if they are not both nil, right?

Focshole commented on mirrorlist_test.go:
```diff
@@ -88,7 +83,7 @@ repos:
 		DownloadTimeout: 200,
 	}
-	if !cmp.Equal(*got, *want) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreUnexported(Repo{})) {
```
same here

gw000 replied:
I've changed these to only ignore the mutex. Let me know if that's all right

Focshole commented:
Thank you for your work!
The idea of moving mutexes to repos seems fine, it is a progress in the proper direction. The issues I see so far are:
- In tests, if you ignore repos pointer, add another check that verifies that the generated Repo structures match (got repos vs want repos)
- Probably the use of RWMutexes was ideal, pacoloco has concurrent issues on all maps so far which needs to be addressed and I'd guess they are the proper tools to fix them. 

Ideally, this would definitely reduce many race conditions but it won't defeat them, as all maps which can be accessed concurrently may show those issues again.

BTW, I am saying those things hadn't tested your work so far, sorry

gw000 commented:
Sure, i can add that test. Can you elaborate on how these errors could still happen? It seems like there are only two maps left (`config.Repos` and `downloadingFiles`), and of those, only `downloadingFiles` could be read/written concurrently, since `config.Repos` isn't written to after it's parsed from YAML.  And `downloadingFiles` is already behind a mutex. So i don't think concurrent map writes are possible anymore, but i could be missing something.

anatol commented:
Thank you for your worj @gw000. I like the idea of using finer-grained concurrency.

gw000 commented:
@Focshole I updated the tests to only ignore the mutex, and check the timestamps. I also added a config test to make sure the new timestamp fields (which are exported to work with `cmp.Equal`) won't be unmarshalled from YAML, if the default keys should happen to be present for some reason.

IMO using RWMutexes would be premature at this point. There would only be one or two very quick reads that could be protected with read locks, and it would add complexity to a critical code path for not much performance gain.

Focshole commented:
> Sure, i can add that test.
I mean, it is enough to add the same checks you see for the whole configs also to the individual structures that should have been created

> Can you elaborate on how these errors could still happen? It seems like there are only two maps left (`config.Repos` and `downloadingFiles`), and of those, only `downloadingFiles` could be read/written concurrently, since `config.Repos` isn't written to after it's parsed from YAML. 


Well, not exactly. The config repos can be updated if it is specified through mirrorlist (it has to read periodically the mirrorlist file if it has been modified, then periodically update the relevant repo urls)

> And `downloadingFiles` is already behind a mutex. So i don't think concurrent map writes are possible anymore, but i could be missing something.

True

Focshole commented on config_test.go:
```diff
@@ -48,16 +48,16 @@ repos:
-	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch")) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch"), cmpopts.IgnoreFields(Repo{}, "mutex")) {
 		t.Errorf("got %v, want %v", *got, *want)
 	}
```
Adding something like

```suggestion
        if *got.Repos != *want.Repos{
	if !cmp.Equal(*got.Repos,*want.Repos){
		t.Errorf("got %v, want %v", *got.Repos, *want.Repos)
	}
	}
```

gw000 replied:
I believe you are. `IgnoreFields` only ignores the fields that are passed as arguments, which are just the mutexes in this case. You can verify that the other fields are checked by changing one of the timestamps in the expected repo in `TestUpdateRepoMirrorlist`.

Focshole replied:
I am dumb, my fault. The check is already here https://github.com/anatol/pacoloco/pull/70/files/4efa1c0b52860cc8c52a931f8daae3a3ee707c9b#diff-4bb821dfaef1884915ee03c5488bded04807301bc3f41fc7001a078c330ae980L63
Sorry for the bother :facepalm:

Focshole commented on mirrorlist.go:
```diff
@@ -10,51 +10,45 @@ import (
-		lastCheck, ok := lastMirrorlistCheck[repo.Mirrorlist]
-		if ok && time.Since(lastCheck) < 5*time.Second {
+		repo.mutex.Lock()
+		defer repo.mutex.Unlock()
```
I'd move the locking after line 29, it is useless IMO to lock it if it won't be modified

gw000 replied:
No, i think we need the lock around the read to prevent `LastMirrorlistCheck` from being written to while it's being read on line 26.

Focshole replied:
You're right, sorry :+1:

Focshole commented on mirrorlist.go:
```diff
@@ -10,51 +10,45 @@ import (
-	if !ok {
-		return fmt.Errorf("repo %v does not exist in config", repoName)
-	}
+func updateRepoMirrorlist(repoName string, repo *Repo) error {
```
locking in this function or on the caller? I have no clue on what is the proper option

Focshole replied:
It makes sense to me! I had separated them when I wrote them because I didn't knew how the code would have evolved

gw000 commented:
Maybe i'm confused but i don't think the race condition you're describing could arise. If the mirrorlist has been updated, the thread that discovers that will lock the repo, preventing any other thread from reading the URLs until the mirrorlist has been parsed and the sources have been update. If, hypothetically, the other thread won the race, then *it* would be the one to find the mirrorlist out of date, and the other thread would wait. Again, i think an RWMutex is very much overkill here.

Focshole commented:
> Maybe i'm confused but i don't think the race condition you're describing could arise. If the mirrorlist has been updated, the thread that discovers that will lock the repo, preventing any other thread from reading the URLs/starting to download until the mirrorlist has been parsed and the sources have been updated. If, hypothetically, the other thread won the race, then _it_ would be the one to find the mirrorlist out of date, and the other thread would wait. Again, i think an RWMutex is very much overkill here.

The logic behind avoiding race conditions involves defining proper access control to the relevant structures that are being accessed concurrently both RW. The structures that I see that are being accessed concurrently are:
```go
Repo.LastMirrorlistCheck
Repo.LastModificationTime
Repo.URLs
```

Thanks to your work, they are much smaller structures than before.
I see that both 
```go
Repo.LastMirrorlistCheck
Repo.LastModificationTime
```
Have been protected with `Repo.mutex`. (BTW, shouldn't it be Repo.Mutex in order to be accessed publicly? Probably I'm wrong here). However `Repo.URLs` access seems unprotected.

In [this](https://github.com/anatol/pacoloco/pull/70#issuecomment-1463531594) scenario,
This instruction:

https://github.com/gw000/pacoloco/blob/fix-map-concurrency/mirrorlist.go#L60

And this instruction:

https://github.com/gw000/pacoloco/blob/fix-map-concurrency/pacoloco.go#L311

Can be executed at the same time.

gw000 commented:
Thanks for explaining, i see the error now. I'll add another mutex.

Focshole commented:
> Thanks for explaining, i see the error now. I'll add another mutex.

well, wait, the original scenario is wrong, but the conclusion that it may occasionally happen is still right. 

With the blocking mutex on `checkAndUpdateMirrorlistRepo`, the only occasion it could happen is if pacoloco has a quite high volume of requests, in which:
1.  the first request does not update the mirrorlist
2. the second request does update the mirrorlist and thus the URLs
3. the first request breaks during that for cycle

Focshole commented on mirrorlist_test.go:
```diff
@@ -83,12 +78,14 @@ repos:
 		DownloadTimeout: 200,
 	}
-	if !cmp.Equal(*got, *want) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Repo{}, "timestampsMutex", "urlsMutex")) {
```
Should check repos

Focshole commented on mirrorlist_test.go:
```diff
@@ -98,7 +95,7 @@ repos:
 	updateMirrorlists()
 	got = config
-	if !cmp.Equal(*got, *want) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Repo{}, "timestampsMutex", "urlsMutex")) {
```
same here

Focshole commented on config_test.go:
```diff
@@ -145,16 +145,16 @@ repos:
 		Prefetch:        &RefreshPeriod{Cron: "0 0 3 * * * *", TTLUnaccessed: 5, TTLUnupdated: 200},
 	}
-	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch")) {
+	if !cmp.Equal(*got, *want, cmpopts.IgnoreFields(Config{}, "Prefetch"), cmpopts.IgnoreFields(Repo{}, "timestampsMutex", "urlsMutex")) {
```
Same here

Focshole reviewed (commented):
The fix is perfect IMO, the only thing it is missing now is the edits on the tests and it is good to go!

Focshole commented:
My fault, sorry for bothering in those tests :facepalm: I tested them and they do what they should. Sorry again

Focshole commented:
It is indifferent to me, if it works it is fine to me! Well done! :+1:
